ipyparallel/cluster/shellcmd.py

```diff
@@ -0,0 +1,866 @@
+
+    def cmd_start(self, start_cmd, env=None, output_file=None):
+        if env:
+            self._log(f"env={env!r}")
+            if not isinstance(env, dict):
+                raise TypeError(f"env must be a dict, got {env!r}")
+
+            # update environment
+            for key, value in env.items():
+                if value is not None and value != '':
+                    # set entry
+                    os.environ[key] = str(value)
+                else:
+                    # unset entry if needed
+                    if key in os.environ:
+                        del os.environ[key]
+
+        if isinstance(start_cmd, str):
+            start_cmd = [start_cmd]
+
+        if not all(isinstance(item, str) for item in start_cmd):
+            raise TypeError(f"Only str in start_cmd allowed ({start_cmd!r})")
+
+        if self.platform == Platform.Windows:
+            # under windows we need to remove embracing double quotes
+            for idx, p in enumerate(start_cmd):
+                if p[0] == '"' and p[-1] == '"':
+                    start_cmd[idx] = p.strip('"')
+
+        self._log(f"start_cmd={start_cmd}  (use_breakaway={self.use_breakaway})")
+
+        if self.platform == Platform.Windows:
+            from subprocess import (
+                CREATE_BREAKAWAY_FROM_JOB,
+                CREATE_NEW_CONSOLE,
+                DEVNULL,
+                Popen,
+            )
+
+            flags = 0
+            if self.use_breakaway:
+                flags |= CREATE_NEW_CONSOLE
+                flags |= CREATE_BREAKAWAY_FROM_JOB
+
+            pkwargs = {
+                'close_fds': True,  # close stdin/stdout/stderr on child
+                'creationflags': flags,
+            }
+            if output_file:
+                fo = open(output_file, "w")
+                pkwargs['stdout'] = fo
+                pkwargs['stderr'] = fo
+                pkwargs['stdin'] = DEVNULL
+
+            self._log(f"Popen(**pkwargs={pkwargs}")
+            p = Popen(start_cmd, **pkwargs)
+            self._log(f"pid={p.pid}")
+
+            print(f'__remote_pid={p.pid}__')
+            sys.stdout.flush()
+            if not self.use_breakaway:
+                self._log("before wait")
+                p.wait()
+                self._log("after wait")
+        else:
+            from subprocess import DEVNULL, Popen
+
+            fo = DEVNULL
+            if output_file:
+                fo = open(output_file, "w")
+
+            p = Popen(
+                start_cmd, start_new_session=True, stdout=fo, stderr=fo, stdin=DEVNULL
+            )
+            print(f'__remote_pid={p.pid}__')
+            sys.stdout.flush()
+
+    def cmd_running(self, pid):
+        self._log(f"Check if pid {pid} is running")
+
+        if self.platform == Platform.Windows:
+            # taken from https://stackoverflow.com/questions/568271/how-to-check-if-there-exists-a-process-with-a-given-pid-in-python
+            import ctypes
```

Comment:
Just from a code organization standpoint, does it make sense to have a `_shellcmd_windows` and `_shellcmd_posix` module with the same API, rather than having all of these imports under `if` statements. This really is two implementations with ~no code shared, so it might be easier to grasp by splitting it to one file per platform (there are only two platforms).

We can keep the class here, and use functions for the platform-dependent implementations. Or a mixin class, if that appeals to you more.

Seeing how these are transmitted, perhaps separate files would complicate things. But at least separate _classes_ seems to make sense to me. What do you think?

Reply:
Makes sense to me. 
In class `ShellCommandReceive` the functions `cmd_start`, `cmd_running ` and `cmd_kill ` are platform dependent. `cmd_mkdir`, `cmd_rmdir`, `cmd_exists` and `cmd_remove` are platform indepedent.  The log and the breakaway flag favours a mixin class concept to me. 
So I would split the code in 3 classes:
`class ShellCommandReceiveBase:` ->platform indepedent code including log functionality
`class ShellCommandReceiveWindows(ShellCommandReceiveBase):`  -> has breakaway flag
`class ShellCommandReceivePosix(ShellCommandReceiveBase):`

And changing the `ShellCommandReceive.__init__(...) ` member function into a factory function `ShellCommandReceive(...)` that returns either a `ShellCommandReceiveWindows` or a `ShellCommandReceivePosix` object. Happy with that solution?

Apart from the `ShellCommandReceive` class there is only `ShellCommandSend._cmd_send` which has plaform dependent code in case the breakaway flag is not supported. I could put this code in a separate member function to have the code a bit better structured. Is that sufficient?

Reply:
That looks sensible to me, thanks.

Reply:
see commit 0eac88a738c21c84dafecbc0d21a9511188ffaa8